packages/utils/src/ethereumHelper.ts

```diff
@@ -19,3 +19,48 @@ export async function signerHelper(
   });
   return signature;
 }
+
+const smartWalletABI = [
+  'function isValidSignature(bytes32 _message, bytes _signature) public view returns (bool)',
+];
+
+enum WalletType {
+  EOA,
+  SMART,
+}
+
+async function getWalletType(
+  address: string,
+  provider: providers.BaseProvider,
+): Promise<WalletType> {
+  const code = await provider.getCode(address);
+  return code === '0x' ? WalletType.EOA : WalletType.SMART;
+}
+
+export async function verifySignature(
+  address: string,
+  message: string,
+  signature: string,
+  provider: providers.BaseProvider,
+): Promise<boolean> {
+  const walletType = await getWalletType(address, provider);
+
+  if (walletType === WalletType.EOA) {
+    const recoveredAddress = utils.verifyMessage(message, signature);
+    return address === recoveredAddress;
+  }
+
+  // Smart wallet
+  const arrayishMessage = utils.toUtf8Bytes(message);
+  const hexMessage = utils.hexlify(arrayishMessage);
+  const hexArray = utils.arrayify(hexMessage);
+  const hashMessage = utils.hashMessage(hexArray);
+
+  const contract = new Contract(address, smartWalletABI, provider);
```

Comment:
This also throws sometimes if the contract doesn't support the ABI, and should be inside the try-catch

Reply:
Are you sure about this ? It is a constructor, so it's sync, so I don't think it can go check the contract code and throw here

Reply:
True. I've had some issues with ethersJS sometimes though. But never mind. Lets go ahead with this